staging/src/k8s.io/apiserver/pkg/server/options/etcd.go

```diff
@@ -223,12 +228,58 @@ func (s *EtcdOptions) Complete(storageObjectCountTracker flowcontrolrequest.Stor
 	}
 
 	if len(s.EncryptionProviderConfigFilepath) != 0 {
-		transformerOverrides, kmsPluginHealthzChecks, err := encryptionconfig.LoadEncryptionConfig(s.EncryptionProviderConfigFilepath, s.EncryptionProviderConfigAutomaticReload, stopCh)
+		ctx, closeTransformers := wait.ContextForChannel(stopCh)
+
+		encryptionConfiguration, err := encryptionconfig.LoadEncryptionConfig(s.EncryptionProviderConfigFilepath, s.EncryptionProviderConfigAutomaticReload, ctx.Done())
 		if err != nil {
+			// in case of error, we want to close partially initialized (if any) transformers
+			closeTransformers()
 			return err
 		}
-		s.transformerOverrides = transformerOverrides
-		s.kmsPluginHealthzChecks = kmsPluginHealthzChecks
+
+		s.kmsPluginHealthzChecks = encryptionConfiguration.HealthChecks
+
+		var kmsPluginHealthzCheck healthz.HealthChecker
+		if len(encryptionConfiguration.HealthChecks) > 0 {
+			kmsPluginHealthzCheck = encryptionConfiguration.HealthChecks[0]
+		}
+
+		s.dynamicTransformers = encryptionconfig.NewDynamicTransformers(encryptionConfiguration.Transformers, kmsPluginHealthzCheck, closeTransformers, encryptionConfiguration.KMSTimeout)
+
+		// enable kms hot reload controller only if the config file is set to be automatically reloaded
+		if s.EncryptionProviderConfigAutomaticReload {
+			// with reload=true we will always have 1 health check
+			if len(encryptionConfiguration.HealthChecks) == 1 {
+				// add post start hook to start hot reload controller
+				// adding this hook here will ensure that it gets configured exactly once
+				err = addPostStartHook(
+					"start-encryption-provider-config-automatic-reload",
+					func(hookContext server.PostStartHookContext) error {
+						// get context
+						// we ignore the cancel func because this context should only be canceled when stopCh is closed
+						ctx, _ := wait.ContextForChannel(hookContext.StopCh)
+
+						kmsConfigController := kmsconfigcontroller.NewDynamicKMSEncryptionConfiguration(
+							"kms-encryption-config",
+							s.EncryptionProviderConfigFilepath,
+							s.dynamicTransformers,
+							encryptionConfiguration.EncryptionFileContentHash,
+							stopCh,
```

Comment:
it's confusing to have stopCh and hookContext.StopCh and ctx (based on hookContext.StopCh), and pass one to NewDynamicKMSEncryptionConfiguration and another to Run...